Add a future annual cycle graph

corviday opened the pull request:
[Adds a present / future comparison graph to single-variable portals showing future values.](https://github.com/pacificclimate/climate-explorer-frontend/issues/145) The difference between January 2010 and January 2040 is _much_ smaller than the difference between January 2010 and June 2010 and gets lost in the noise. So in addition to the usual annual cycle time series for each climatology, there is a row of bar charts along the bottom, showing the difference between each climatology and the present-day base climatology. This lets a user see, in more detail, "Oh, it will be warmer in January but colder in March according to this model."

![screenshot from 2018-04-17 14-05-44](https://user-images.githubusercontent.com/4512605/38896893-d2cb47d2-4248-11e8-85e5-9fa0e4d0c0f8.png)

This PR is much smaller than it looks. Most of the putatively new or deleted lines actually stem from splitting the over-long `chart.js` into three files, `chart-generators.js`, `chart-formatters.js`, and `chart-transformers.js`, which was straightforward and didn't require rewriting anything beyond file paths. 

Actual new code is the following:

- `chart-formatters.js`: `padYAxis()` and `hideSeriesInTooltip()`
- `chart-transformers.js`: `makeAnomalyGraph()` and `addAnomalyTooltipFormatter()`
- `AnomalyAnnualCycleGraph.js`
- `SingleDataController.js` : the section in render() that calls AnomalyAnnualCycleGraph
- `chart-formatter-tests.js` : tests for `padYAxis()` and `hideSeriesInTooltip()`
- `chart-transformer-tests.js`: tests for `makeAnomalyGraph()` and `addAnomalyTooltipFormatter()`

Everything else is from splitting up chart.js, splitting up chart.js's tests, or editing imports to point to where chart functions ended up.

rod-glover commented:
A few questions that might have been resolved already in discussions I wasn't part of:

1. Why bar charts (as opposed to line graphs)?
2. How many climatologies might be shown on this graph, and thus in the difference display at the bottom? If more than 2 or 3, this could get awkward for bar charts. Possibly, but not certainly, a set of line graphs might be easier to read in those cases.
3. Is the vertical space allocated for the difference display large enough? It seems like it might be hard to read the differences very accurately.

rod-glover commented on src/core/chart-formatters.js:
```diff
@@ -0,0 +1,284 @@
+ *   padding - the amount of extra y-axis space to add, expressed as a 
+ *             multiple of the existing data span. 
+ */
+var padYAxis = function (graph, axis = "y", direction = "top", padding = 1) {
```
Not sure where this habit of declaring functions as variables got started, but it's unnecessary and in some environments can result in anonymous functions where you really want a named function. (I think Babel is smart about naming functions, but some environments aren't and these kinds of anonymous functions make debugging a lot harder.) 

Much prefer a named function, i.e.,
```
function padYAxis(...) { ... }
```

rod-glover replied:
On further reflection, I'm not sure if fixing this issue throughout belongs in this PR. It's always tempting to slip in some code improvement like this, but it does add unnecessary changes.

rod-glover commented on src/core/chart-formatters.js:
```diff
@@ -0,0 +1,284 @@
+    if(predicate(graph.data.columns[i])) {
+      hidden.push(graph.data.columns[i][0]);
+    }
+  }
```
This could be simplified as
```
const hidden = _.pluck(_.filter(graph.data.columns, predicate), 0);
```

corviday replied:
Ooo, nice.

rod-glover commented on src/core/chart-formatters.js:
```diff
@@ -0,0 +1,284 @@
+        const seriesMin = _.min(series);
+        min = seriesMin < min ? seriesMin : min; 
+      }
+    }
```
This loop could be eliminated and simplified as:
```
min = _.min(
   _.map(
      _.filter(graph.data.columns, series => series[0] !== 'x' && graph.data.axes[series[0]] === axis),
      series => _.min(series)
   )
);
```

Better with chaining:
```
_.chain(graph.data.columns)
.filter(graph.data.columns, series => series[0] !== 'x' && graph.data.axes[series[0]] === axis)
.map(_.min)
.min()
.value();
```

Alternatively:
```
_.chain(graph.data.columns)
.filter(graph.data.columns, series => series[0] !== 'x' && graph.data.axes[series[0]] === axis)
.flatten()
.min()
.value();
```

rod-glover commented on src/core/chart-formatters.js:
```diff
@@ -0,0 +1,284 @@
+        const seriesMax = _.max(series);
+        max = seriesMax > max ? seriesMax : max; 
+      }
+    }
```
Similarly

rod-glover replied:
And you can factor out the `series` predicate that is common to both computations. That would make the `underscore` expressions tighter and probably possible to put on one line with clarity.

corviday commented:
[Here's a docker for this branch](http://docker-dev01.pcic.uvic.ca:30080/#/). 

I'll take all the UI feedback I can get. :)

rod-glover commented on src/core/chart-transformers.js:
```diff
@@ -0,0 +1,283 @@
+  
+  // function that determines whether a data series an anomaly series.
+  // used to format anomalies differently than data
+  let isAnomaly = function (series) {
```
Please use just a function declaration: `function isAnomaly(series) { ... }`.

Also, `const` here would be the appropriate declaration; `let` is for values that will be reassigned, i.e., that are variable and that does not seem to be your intention here.

rod-glover commented on src/core/chart-transformers.js:
```diff
@@ -0,0 +1,283 @@
+  
+  // classifier function that matches each "anomaly" data series with
+  // the nominal series it is based on. Used to match colours.
+  let anomalyMatcher = function(series) {
```
Ditto `function`

rod-glover commented on src/core/chart-transformers.js:
```diff
@@ -0,0 +1,283 @@
+  // classifier function that matches each "anomaly" data series with
+  // the nominal series it is based on. Used to match colours.
+  let anomalyMatcher = function(series) {
+    let sName = series[0];
```
`const`

rod-glover commented on src/core/chart-transformers.js:
```diff
@@ -0,0 +1,283 @@
+  
+  // ranking function that assigns anomaly series lower results than 
+  // nominal series, used to make them distinguishable.
+  let anomalyRanker = function (series) {
```
Ditto `function`

rod-glover commented:
More thoughts on UI:

1. In the difference display, I'd eliminate the means, which by definition always have difference value zero.
2. I'd definitely allocate more space to the difference display; at least twice what is there now, possibly much more. (I think, but am not sure, that this display is roughly equally important to the absolute value display above it. Thoughts?)
3. I think I still prefer the line graph for differences, but let's compare the options with a larger area for the difference display.

corviday commented:
I've redone [the docker](http://docker-dev01.pcic.uvic.ca:30080/#/) for this branch and changed the graph formatting to make the difference graph and the absolute value about the same size.

As a way of making it more obvious which axis goes with which graph, I also experimented with blanking out some of the axis labels.

Opinions?

rod-glover commented on src/core/chart-formatters.js:
```diff
@@ -0,0 +1,328 @@
+  
+  if(_.isUndefined(min)) {
+    min = _.min(_.map(axisSeries, series => _.min(series)));
+  }    
```
👍

rod-glover commented on src/core/chart-formatters.js:
```diff
@@ -0,0 +1,328 @@
+  let max = graph.axis[axis].max
+  const axisSeries = _.filter(graph.data.columns, series => {
+      return series[0] !== 'x' && graph.data.axes[series[0]] === axis
+    });
```
Since this code is repeated verbatim at least one place elsewhere, I'd pull it out into a common function. Possibly that function would take the axis identifier ('x') as a parameter.

rod-glover replied:
No?

rod-glover commented on src/core/chart-transformers.js:
```diff
@@ -0,0 +1,287 @@
+  }
+  
+  const xseries = _.filter(graph.data.columns, series => seriesNameContains(series, x));
+  const yseries = _.filter(graph.data.columns, series => seriesNameContains(series, y));
```
👍

rod-glover commented on src/core/chart-transformers.js:
```diff
@@ -0,0 +1,287 @@
+      c3Data.columns[0].push((tuples[i][0] + tuples[i+1][0])/2);
+      c3Data.columns[1].push(null);
+    }
+  }
```
Wherever you find yourself iterating through an array and pushing items onto another array derived from the iterated array, you can usually reformulate it more clearly as a set of filter, map, and other operations (courtesy of underscore). 

When the set of filter and map operations gets larger that 2 or 3 (or even 1!), you can use chaining (part of underscore) to make it clearer: Chaining means that the array arguments you'd have to pass explicitly to the filter, map, etc. functions are omitted because the operations are methods of the array (wrapped as a chainable object by underscore). It's very nice, and close to the syntax of some of the more delightful functional languages.

rod-glover replied:
I see your point; I didn't quite notice the double pushes. Just for reference, the double pushes could be accommodated by mapping to pairs (two-element arrays) and then flattening the result. (In this case, you'd also need to push the final two results afterwards. That plus other minor complications makes my functional solution sound worse than your iterative solution in this case.)

I don't see your reduce solution here, but it doesn't _sound_ like a better solution.

corviday replied:
Sorry about that; I thought github would update with what I'd changed this section to. [Here's the reduce solution.](https://github.com/pacificclimate/climate-explorer-frontend/pull/146/files#diff-dc15f875105ec95e990b237fac511880R94) I wasn't terribly happy with it.

rod-glover commented:
Lee, nice work. 

UI thoughts:

1. The larger space allocated for the difference display is much better to my eye. Perhaps the team should decide what fraction; I kinda like the size it is now, but it could be smaller and still have the same impact I think.
2. I like the blanked axis labels, they make things clearer. As an alternative, and only just thinking aloud here, could there be two different graphs, with the x-axis of one somehow omitted or suppressed. That might simplify the label juggling if it's possible.
3. In the difference display, can the baseline values (with constant difference zero) be removed?
4. I noticed that on my laptop screen, the layout doesn't work very well: the map overlaps the graphs. Probably just an easy Bootstrap fix.

corviday commented:
2. I did them as one graph so they'd share the tooltip, which gives both the absolute and difference values, and provides a line connecting the absolute and difference graphs. I think the tooltip is pretty great, but I'm open to doing two separate graphs if we prefer.

3. Yes, definitely. I personally found the difference graph is easier to understand with the zero-difference line providing an ersatz axis, but at this point, I've reworked this graph enough I'm not sure my intuitions about it are useful. :) 

4. This is issue https://github.com/pacificclimate/climate-explorer-frontend/issues/70 

I'll tidy the code and get up a version without the zero-difference line Monday.

rod-glover commented:
2. Ah, the shared tooltip is a good idea. Probably worth retaining.
3. I get the value of the zero axis. Could it be achieved with just two points and one line segment joining them?

corviday commented:
[Docker for this branch](http://docker-dev01.pcic.uvic.ca:30080/#/) has been updated and is now caught up to all the UI suggestions! Removed the "anomaly" line for the 2010 climatology (all zeroes) and switched the bar charts to line charts.

rod-glover commented on src/core/chart-formatters.js:
```diff
@@ -0,0 +1,328 @@
+      if(predicate(column)){
+        hiddenSeries.push(seriesName);
+      }
+    }
```
Ditto comments re. `_.each` vs. for-of, and `!==` vs. `isEqual` for simple values.
Also, the nested if statements should be collapsed to one.

FWIW, this could be written as
```
const hiddenSeries = 
   _.chain(graph.data.columns)
   .filter(column => column[0] !== 'x' && predicate(column))
   .map(column => column[0]);
```
which depending on reader might be deemed easier to read.

rod-glover commented:
I like the new Future Anomaly graph better.

The y-ranges for value and difference graphs overlap; can they be made entirely disjoint? The overlap is confusing, especially when (as I see in the default model, var, emission), the first difference actually extends into the range of the value graph (looks like -20 C). It would be better to have some blank space between the two graphs.

rod-glover commented:
Nice work, Lee.
